### ISSUE TITLE
Itch.io service

### ISSUE THREAD

**Issue by RobLoach** [lutris/lutris#243]
http://itch.io/app

Does having an [Itch.io](http://itch.io/) runner for Lutris make sense? I don't see any available command line parsing in [the source](https://github.com/itchio/itch). We might have to add them.

**Comment by strycore** [comment]
A runner for itch.io would not make sense, they're a platform providing Linux or Windows games. But it does make sense to support it in the same way we will support GOG and Humble Bundle. It's a very good thing they have an open source client, it will make it much easier to integrate their API to our client.

**Comment by ProfessorKaos64** [comment]
I hope this comes to fruition as well, would love to see this.

**Comment by RobLoach** [comment]
```
itch itchio://install/GAMEID
```

And eventually, `itch itchio://launch/GAMEID`.

[itch launch command line arguments](/itchio/itch/issues/324)

**Comment by ProfessorKaos64** [comment]
hazah :)

**Comment by strycore** [comment]
I was thinking of doing more of a complete replacement for itch.io's client with something that would handle users authentication and fetching their library. This is what is planned for Humble Bundle and GOG so we'll probably do the same for Itch.io.

**Comment by RobLoach** [comment]
How would this work? Would we reimplement some of their API to download install and play games? Using the client as a base would save quite a bit of time.

**Comment by strycore** [comment]
It will work the same way as Humble Bundle and GOG, there is an API, we call it and we use it to integrate the platform with Lutris. I have no idea how we could provide any kind of integration if we used the official client.

**Comment by RobLoach** [comment]
I assumed it would be similar to how we use the Steam Runner. Invoke Itch to install and run games just like how it works with the Steam client.

**Comment by strycore** [comment]
Yeah, it's a bit different. We already have installer for a bunch of Itch.io, GOG and Humble games because the games are provided DRM free in easy to process formats (zips, tar.gzs or, worst case, Windows installers). In the case of Steam, there is not way of getting the games without the Steam client, which is why we have to rely on it.

**Comment by Kabouik** [comment]
Just a shy bump to check if this is still on the roadmap. Having most platforms supported in Lutris would really help keeping a tidy library.

I regularly remember about games I own but that are not installed or listed in any of my clients, and have to look through old emails to find them again. However, most frequently, I'm sure I just totally forget about them and virtually lose them. One may argue that if I can forget them, then I don't need them anymore, and it's partly true, but sometimes they are really good games that I enjoyed but stopped playing for a while, and would be glad to play them again when I have time. It's just too difficult to remember everything when more than 100-150 games accumulated over the years, and Lutris could really help in that regard by expanding to additional platforms.

**Comment by strycore** [comment]
It's still planned :) 

Once we ship GOG support, it will be easier to add support for additional platform. But note that since Itch.io already has a Linux client, we might go for Humble Bundle first.

**Comment by Kabouik** [comment]
Good news, thanks! I can't wait, and am eagerly looking forward to the GOG support as well.

**Comment by Kabouik** [comment]
So this is not on the roadmap anymore? Itch.io is a good client, but one of the key features of Lutris is to centralize. If good clients are not integrated, then we'll be left with one of the issues Lutris is aiming to solve: having multiple clients and libraries in parallel.

**Comment by lionirdeadman** [comment]
AFAIK Itch.io hasn't worked (although it is apparently planned since 2016 : https://github.com/itchio/itch/issues/433) to have any kind of wine layer so this makes all of their windows-only games unavailable to us. I trust that they'll work towards it but it seems to have quite a low priority (which is fair). So a runner might still be something we'd want. Or at the very least, the option to launch itch.io games from Lutris. What do you think @strycore ?

**Comment by strycore** [comment]
If it's a matter of having Wine compatible games then the first step is to have a Lutris installer like we have for Orgin and Uplay

**Comment by lionirdeadman** [comment]
I'll give it a shot when the Lutris flatpak gets better then, I'll save this issue.

**Comment by fasterthanlime** [comment]
Just a quick note, official Wine support in https://itch.io/app is still planned, I've just been working (again) on some fundamental changes to the app/butler's infrastructure, and also dealing with personal health.

A Lutris/itch app integration could be nice (like Playnite did), but I'd recommend waiting a bit as I'm changing the RPC protocol used to communicate with butler (which Lutris would have to implement) - and I'd rather have only Playnite to guide through that rather than two clients.

Of course if you don't want to wait that's fine too! It's all out there anyway.

**Comment by PopeRigby** [comment]
@fasterthanlime How's the progress on the RPC protocol? Should this issue be reopened?

**Comment by fasterthanlime** [comment]
> How's the progress on the RPC protocol? Should this issue be reopened?

I've kept the old protocol and just today added preliminary Wine support - I'm also planning on supporting external launches (the GameHub folks requested this, might come in handy for Lutris too).

If anyone wants to get started with http://docs.itch.ovh/butlerd/master/#/ - I'm sure our paths will cross in the near future. I'm not aware of a butler JSON-RPC client in Python yet, so maybe start there? Let me know if you need any help!

**Comment by hoxu** [comment]
@strycore would you consider reopening this issue?

I think it would be useful to import the Itch client library to Lutris, similar to how it's possible to import Steam games.

**Comment by strycore** [comment]
Itch support is still planned regardless of the status of this ticket.

**Comment by kahveciderin** [comment]
What is the current progress on this? Would love to work on this and submit a pr, if it isn't already implemented.

**Comment by strycore** [comment]
> What is the current progress on this? Would love to work on this and submit a pr, if it isn't already implemented.

Please! I'm having problems figuring out the authentication for Itch so it you manage to get this working, it would be super helpful!!

**Comment by judemille** [comment]
> Please! I'm having problems figuring out the authentication for Itch so it you manage to get this working, it would be super helpful!!

What's your problem figuring out auth? I'm not familiar with the app or API but I may be able to crack the code.

**Comment by strycore** [comment]
I tried to understand how to get an auth token like we do for the other services but I got really confused with the existing APIs and tools. It's been a while since I've taken a shot at Itch.io so I don't remember all the details, I remember the auth wasn't handled by the client itself but another CLI tool.

**Comment by merwok** [comment]
For an app like lutris (opposed to a game), I think it’s OAuth: https://itch.io/docs/api/oauth

**Comment by strycore** [comment]
> For an app like lutris (opposed to a game), I think it’s OAuth: https://itch.io/docs/api/oauth

Either I missed that when I was experimenting with Itch or it's new. If we can use OAuth, then the integration should be fairly simple.

**Comment by hagabaka** [comment]
I think OAuth is for Itch's [Serverside API](https://itch.io/docs/api/serverside) which allows querying information about your games, but can't help with downloading/running games. Lutris would need to use the Butler API as @fasterthanlime pointed [earlier](https://github.com/lutris/lutris/issues/243#issuecomment-548087555). Basically Butler runs as a daemon, and an app like Lutris can communicate with it through JSON-RPC, to make it authenticate, fetch game information, download games, and launch them.

**Comment by abdonkov** [comment]
Maybe you could look how the Playnite developers have done the [integration with itch.io](https://github.com/JosefNemec/PlayniteExtensions/tree/master/source/Libraries/ItchioLibrary) for some inspiration.
But it's basically what @hagabaka said. Running Butler and communicating with it through JSON-RPC.

**Comment by relma2** [comment]
Im gonna bump this feature request. Itch is gaining more and more steam (no pun intended) and kitch is really just their web browser page with widgets. Lutris integration would go a long way.

**Comment by mossymossmoss** [comment]
As @relma2 said, Itch is one of the biggest Indie platforms out there, it makes sense that Lutris would support it. I have over 5000 games on it, so maybe I am a bit partial (yeah it was from some of those massive bundles lol). That being said, the ability to add them easily without the need of manually downloading and setting them up would be great

**Comment by cheater** [comment]
note with those megapacks that everyone bought at some point it would make sense to have some way to differentiate between, um, you know, games you might want to have, vs, eh, games you'd be less excited about. trying to be very neutral in my wording here ...

**Comment by Shadow53** [comment]
> note with those megapacks that everyone bought at some point it would make sense to have some way to differentiate between, um, you know, games you might want to have, vs, eh, games you'd be less excited about. trying to be very neutral in my wording here ...
> 
> and that without every user having to go through a list of 5000 games and having to choose "shovelware" or "not shovelware" on each one of them

IIRC, with at least one of those bundles I had add games one by one into my library before they showed up under "Owned Items". I imagine the same would apply here, that big bundle items would not show up unless already added to the user's library.

**Comment by JacobTravers** [comment]
> IIRC, with at least one of those bundles I had add games one by one into my library before they showed up under "Owned Items". I imagine the same would apply here, that big bundle items would not show up unless already added to the user's library.

This is how it worked for me on a recent bundle purchase over the summer.

Also gamehub seems to have their itch integration working: https://github.com/tkashkin/GameHub/tree/master/src/data/sources/itch

**Comment by GoGoOtaku** [comment]
I'm about 50% done implementing this.
I will probably not finish it this year (I work rather slow due to personal reasons) but I can already populate the games list and technically have it the fundamentals for installing and launching apps. I am mostly busy figuring out how Lutris works internally to populate stuff into the UI.
Like electron itch and gamehub I use their helper app called butler. Itch doesn't have a nice API and letting butler take care of everything is rather nice. For that purpose I am currently implementing it similar to steam aka having a service, a runner and maybe a custom installer like `steam_installer.py` if it can't be helped. As soon as I can install, launch and uninstall games I will open a WIP merge request to get some feedback. Stuff like login through the UI I would do after.

I am mostly "announcing" this here as I feel like I invested enough time in this now where I feel kinda iffy if I would find out tomorrow that it was all for naught.

> > note with those megapacks that everyone bought at some point it would make sense to have some way to differentiate between, um, you know, games you might want to have, vs, eh, games you'd be less excited about. trying to be very neutral in my wording here ...
> > and that without every user having to go through a list of 5000 games and having to choose "shovelware" or "not shovelware" on each one of them
> 
> IIRC, with at least one of those bundles I had add games one by one into my library before they showed up under "Owned Items". I imagine the same would apply here, that big bundle items would not show up unless already added to the user's library.

There is currently no (easy and safe) way to access games purchased in those bundles until their manually added. So yes this will very likely be this way here. The only way I see this being done is crawling through their actual website which would neither be fun to code, nor stable.

**Comment by danieljohnson2** [comment]
I can't speak for @strycore, so I will anyway.

I feel sure that an itch.io service would be a very welcome PR.

Personally I hope you'll stick around and fix the inevitable bug reports for a bit afterwards, but that does not seem to be a requirement around here.

**Comment by strycore** [comment]
That's great news! 
And on top of that, I'll make your life a lot simpler!
You only need to implement a service, nothing else. You don't need a runner and you don't need a custom installer like in steam_installer.py.
Steam has those because Lutris interacts with the client. In the case of Itch, Lutris is the client. The runners used should be either Wine or Linux depending on the game. It's closer to Humble Bundle in that regard.

Note that we have ways to get entire store libraries for GOG and Steam. Steam has an API endpoint containing all games but for GOG, we do have a scraper so it's not entirely out of the question for Itch. Note that the scraper would run on a server and not the user's machine. The data from the scraper gets sent to the Lutris website and the data is integrated in our API.

**Comment by GoGoOtaku** [comment]
I do communicate with a "client" but in hindsight it is probably best to not run the games through it  (itch.io games are all DRM free anyway afaik).
I was using the runner to install the client but after looking into it again I might just have the client as a game (like origin).
The client fetches data, installs and uninstalls games, alternatively download the zipped build, but most importantly handles the login. It would be a shame to not use it since it abstracts pretty much everything into a nice to work with package, which should increase maintainability. The communication is json over a socket. Since the client is a tiny go app it has no dependencies but the kernel itself (aka it is statically linked)

**Comment by strycore** [comment]
The case of butler is a special one as it's neither a runner nor a game, this just an abstraction over the Itch API, which we don't normally have since we're doing the HTTP requests ourselves.

I think the best suited place for butler is as a runtime component. If you can send a link to a build, I'll make it available through our API and you'll be able to use it in your PR.

**Comment by GoGoOtaku** [comment]
I looked into the executable again and before we do anything with butler I would like to look at something again that might make this like super simple.

**Comment by mindmeshlink** [comment]
I'd like to note that I definitely would really appreciate an implementation that doesn't need to use the official Itch client in any capacity, if it's possible. If I wanted to run the Itch client on my computer, I wouldn't be using Lutris to launch my games in the first place.

**Comment by GoGoOtaku** [comment]
> I'd like to note that I definitely would really appreciate an implementation that doesn't need to use the official Itch client in any capacity, if it's possible. If I wanted to run the Itch client on my computer, I wouldn't be using Lutris to launch my games in the first place.

This wasn't about the itch client but a small command-line utility that would have worked in the background. The main reasons why I started work on this is because the itch client is based on electron and I try to keep electron to discord (which I have no choice for). I already used Lutris for GOG (because GOG Galaxy doesn't have a native client) and all other clients are Java based which I also rather not want. So no, no matter what, I will not hand in anything using the official itch client.

**Comment by GoGoOtaku** [comment]
Update: I "extracted" the API out of the public go repositories and have now managed to get my library, download an archive and install a Linux game. Right now I have to work on why it doesn't want to download my other test game (which would be wine). There is still a lot to do tho. A lot of this is based on fairly case specific scenarios (game distributed in just a single archive, no extras or asset files, etc).

**Comment by mossymossmoss** [comment]
@GoGoOtaku would it be possible to have an option to draw your itch.io catalogue from a collection rather than your library? I ask this for two reasons, first as itch.io has many free games which you can't add to your library, only to collections, and thus you wouldn't be able to install/update them via lutris. Secondly, a more personal reason, I happen to have over 1600 games in my itch library which I got from various bundles, and that causes a very large amount of lag when opening in Lutris, to have a designated collection of games  I want to view would be faster for me :sweat_smile:

**Comment by GoGoOtaku** [comment]
This is actually something I thought about a lot. Having such a massive library is partially why big bundles aren't automatically added to the library. The big issue is that there is currently no way of having service specific settings.
Giving people access to their purchases seemed like the most important thing and is how virtually all other (online) services work.
Further you can always add free games via scripts. I added HoloCure to my library like that and update it via the main library just fine ( https://lutris.net/games/holocure/ ). The biggest barrier here is to get the game id for writing the script which I did via the API by manually opening [`https://api.itch.io/search/games?query="holocure"`](https://api.itch.io/search/games?query="holocure") there might be some normal way to access that.

So in the end I came to the conclusion that this would cover most use cases and would be the best way for an mvp; which is what I was aiming for to keep it simple.

Is it technically possible to use collections as source tho? Yes! I actually wanted to make new issues to discuss implementation of some missing stuff (see #4690) with the maintainers but... uhm... I forgor. Let me do that actually.

**Comment by mossymossmoss** [comment]
Before service specific settings are added, you could check the user's itch account for a collection called "Lutris" or something, and if it doesn't exist, just default back to the purchases. It wouldn't be easy to make it clear that it's an option, perhaps some sort of popup like is currently used when the warning is given about signing into Humble could work, but to have the functionality there in would be great

**Comment by GoGoOtaku** [comment]
I thought about something similar. I'd prefer to continue discussing this in the new issue.
